quantumblacklabs/kedro-viz
# [KED-2517] Contracting modular pipelines

## Description

Implements modular pipelines nodes on the flowchart component, that represent multiple elements of a modular pipeline in a single node when collapsed using the sidebar tree controls.

<img width="1131" alt="Screenshot 2021-08-25 at 12 18 14" src="https://user-images.githubusercontent.com/45365769/130781468-ab0909a0-e7d5-46dc-9819-435d0912fcec.png">

## Development notes

The majority of this work was originally implemented by Richard, but required some further work to complete that I have implemented here.

### Caveats
This work is specifically front-end based, since the user's filters can change the graph dynamically at any time, including forming graphs that are unnatural by Kedro standards, so implementation has to be more general and robust than might be expected.

### Algorithm
This PR uses an [edge contraction](https://en.wikipedia.org/wiki/Edge_contraction) based algorithm to merge adjacent nodes together resulting in a simplified final graph.

### Handling disconnected graph components
If a modular pipeline has disconnected components, then this will naturally result in two disconnected modular pipeline nodes.

### Handling external graph components
In the case that a modular pipeline links to an element external to the modular pipeline, such as a dataset from a different namespace, these elements can not merge to a single node. This naturally forms a split of the modular pipeline over two or more modular pipeline nodes, with externals inbetween.

### Handling graph cycles
In particular it's possible that an external component will form a cycle if it links back into the same modular pipeline, a practical example of which can be seen in the shuttles demo dataset, but can be specifically seen in this new minimal [test dataset](http://localhost:4141/?data=contracting).

To handle this there is a special rule ([code](https://github.com/quantumblacklabs/kedro-viz/pull/538/files#diff-7488e78efcbb12d967ddb953ce65d912fec7cc5dd0d7ac4bd3763d7f9782b8c2R131-R154)) introduced to avoid introducing cycles when contracting these cases which would otherwise not form a valid DAG as required, which could break other features such as the layout engine. 

The approach to handle them was discussed with @idanov  and @limdauto and concluded that choosing a preferred edge direction and dropping the cyclic edge was an acceptable solution when following Kedro logic.

## QA notes

A new minimal [test dataset](http://localhost:4141/?data=contracting) was added that includes a potential cycle when the modular pipeline is contracted.

**Note:** I've updated some of the existing unit tests but there are still some more to do as it stands. I've also added some tests to cover the new feature but more might be needed (especially around the sidebar UI interaction).

To manually test:

- open the browser console to check for any errors or warnings
- open a datasets `animals` and `demo` (that have modular pipeline nodes)
- on the sidebar toggle various modular pipelines open / closed
- also try toggling sub modular pipelines at various depths
- check that the output graph has contracted the closed modular pipelines in a reasonable way
- try in combination with various filters on the sidebar
- try in combination with layers enabled

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
